tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp

```diff
@@ -0,0 +1,596 @@
+
+    EXPECT_CALL(mock_dispatcher, dispatch_call_result(_)).WillOnce(Invoke([](const json& call_result) {
+        auto response = call_result[ocpp::CALLRESULT_PAYLOAD].get<ReserveNowResponse>();
+        EXPECT_EQ(response.status, ReserveNowStatusEnum::Accepted);
+    }));
+
+    this->reservation->handle_message(request);
+}
+
+TEST_F(ReservationTest, handle_reserve_now_all_connectors_not_available) {
+    // Try to make a reservation with all connectors unavailable. Since the evse manager has the last word in this,
+    // we try to do the request and it can be accepted anyway (or at least the correct reason for not accepting the
+    // reservation can be returned, if this is a real scenario).
+    const ocpp::EnhancedMessage<MessageType> request =
+        create_example_reserve_now_request(std::nullopt, ConnectorEnum::cTesla);
+
+    EvseMock& m1 = evse_manager.get_mock(1);
+    EXPECT_CALL(m1, does_connector_exist(ConnectorEnum::cTesla)).WillOnce(Return(true));
+
+    EvseMock& m2 = evse_manager.get_mock(2);
+    EXPECT_CALL(m2, does_connector_exist(ConnectorEnum::cTesla)).Times(0);
+
+    EXPECT_CALL(reserve_now_callback_mock, Call(_)).WillOnce(Return(ReserveNowStatusEnum::Accepted));
+
+    EXPECT_CALL(mock_dispatcher, dispatch_call_result(_)).WillOnce(Invoke([](const json& call_result) {
+        auto response = call_result[ocpp::CALLRESULT_PAYLOAD].get<ReserveNowResponse>();
+        EXPECT_EQ(response.status, ReserveNowStatusEnum::Accepted);
+    }));
+
+    this->reservation->handle_message(request);
+}
+
+TEST_F(ReservationTest, handle_reserve_now_non_specific_evse_successful) {
+    // Try to make a non evse specific reservation which is accepted.
+    const ocpp::EnhancedMessage<MessageType> request =
+        create_example_reserve_now_request(std::nullopt, ConnectorEnum::cTesla);
+
+    EvseMock& m1 = evse_manager.get_mock(1);
+    EXPECT_CALL(m1, does_connector_exist(ConnectorEnum::cTesla)).WillOnce(Return(true));
+
+    EXPECT_CALL(reserve_now_callback_mock, Call(_)).WillOnce(Return(ReserveNowStatusEnum::Accepted));
+
+    EXPECT_CALL(mock_dispatcher, dispatch_call_result(_)).WillOnce(Invoke([](const json& call_result) {
+        auto response = call_result[ocpp::CALLRESULT_PAYLOAD].get<ReserveNowResponse>();
+        EXPECT_EQ(response.status, ReserveNowStatusEnum::Accepted);
+    }));
+
+    this->reservation->handle_message(request);
+}
+
+TEST_F(ReservationTest, handle_reserve_now_specific_evse_successful) {
+    // Try to make a reservation for an existing evse, which is accepted.
+    std::optional<ConnectorEnum> tesla_connector_type = ConnectorEnum::cTesla;
+
+    const ocpp::EnhancedMessage<MessageType> request = create_example_reserve_now_request(2, ConnectorEnum::cTesla);
+
+    EvseMock& m1 = evse_manager.get_mock(1);
+    EXPECT_CALL(m1, does_connector_exist(ConnectorEnum::cTesla)).Times(0);
+    EXPECT_CALL(m1, get_connector_status(tesla_connector_type)).Times(0);
```

Comment:
Same here, we don't really care if the code calls this right? Would be weird if it does but it should not fail this as long as the wanted behavior is achieved.

Maybe there is another way to specifically test that it does select the right evse?

Reply:
Ok you are right. This is harder to maintain. I removed it. I don't see how I can test if the reservation was placed for a specific evse (and not for the other one). But we should have other test that do that.